### PR TITLE
Adding random values to IR

### PR DESCRIPTION
**Pull request by Overcastan**
Partly addressing #86.
- Added support for random values in IR.
- Added indexed access to random values by custom name.
- Added integration test for random values.

### CODE REVIEW

File: air-script/tests/aux_trace/aux_trace.air [hunk 1]
```diff
@@ -14,15 +14,15 @@ boundary_constraints:
     enf p0.first = 1
     enf p0.last = 1
 
-    # auxiliary boundary constraint with a random value
-    enf p1.first = $rand[0]
+    # auxiliary boundary constraint
+    enf p1.first = 1
```
Comment (grjte) [inline air-script/tests/aux_trace/aux_trace.air:18]: I said the wrong thing in my previous review, sorry. I had misunderstood that we're removing the default `$rand` option and making the declaration required.

Given that, I'd like to keep the random values in this test, so instead of removing these, we should add the `random_values` declaration (as you did the first time). Sorry for the mistake!

File: air-script/tests/random_values/random_values.air [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+random_values:
+    rand: [x, y[14], z]
+
+integrity_constraints:
+    enf c' = z - $rand[0] + y[2]
+
+boundary_constraints:
+    enf c.first = $rand[5] + y[2] + z
+    enf c.last = x + $rand[15] + y[10]
```
Comment (grjte) [inline air-script/tests/random_values/random_values.air:18]: This test looks great.

nit: let's put the `boundary_constraints` before the `integrity_constraints`, since we usually use that order
nit 2: maybe we should use a different name instead of `rand`, like `alphas`, since these tests also help serve as examples/documentation of functionality.

File: air-script/tests/main.rs [hunk 3]
```diff
@@ -109,3 +109,20 @@ fn indexed_trace_access() {
...
+    let expected = expect_file!["random_values/random_values.rs"];
+    expected.assert_eq(&generated_air);
+
+    let generated_air = Test::new("tests/random_values/random_values_simple.air".to_string())
+        .transpile()
+        .unwrap();
+
+    let expected = expect_file!["random_values/random_values.rs"];
+    expected.assert_eq(&generated_air);
+}
```
Comment (grjte) [inline air-script/tests/main.rs:128]: small nit: I would probably prefer to test the simpler case first (`random_values_simple.air`), and then to give the more complex test a more specific name, like `random_values_bindings.air` or `random_values_named.air` and put it second.

File: air-script/tests/random_values/random_values_simple.air [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+random_values:
+    rand: [16]
+
+integrity_constraints:
+    enf c' = $rand[15] - $rand[0] + $rand[3]
+
+boundary_constraints:
+    enf c.first = $rand[5] + $rand[3] + $rand[15]
+    enf c.last = $rand[0] + $rand[15] + $rand[11]
```
Comment (grjte) [inline air-script/tests/random_values/random_values_simple.air:18]: nit: let's swap the order of the `boundary_constraints` and `integrity_constraints`

File: ir/src/symbol_table.rs [hunk 5]
```diff
@@ -151,6 +159,25 @@ impl SymbolTable {
         Ok(())
     }
 
+    /// Adds all random values by their identifier names and array length.
+    pub(super) fn insert_random_values(
+        &mut self,
+        values: &RandomValues,
+    ) -> Result<(), SemanticError> {
+        // TODO: add name of random values vector to the identifiers. For now it can be called only
+        // by `$rand`
```
Comment (grjte) [inline ir/src/symbol_table.rs:168]: Is this left out of this PR because it's quite complicated to do? If we're not including it here, can you create a sub-issue for it so we don't lose track of it?

File: ir/src/symbol_table.rs [hunk 6]
```diff
@@ -249,6 +280,16 @@ impl SymbolTable {
                     ))
                 }
             }
+            IdentifierType::RandomValue(_offset, size) => {
```
Comment (grjte) [inline ir/src/symbol_table.rs:283]: nit: it's more idiomatic to use `_` instead of `_offset`

File: ir/src/constraints/graph.rs [hunk 7]
```diff
@@ -108,13 +108,18 @@ impl AlgebraicGraph {
                 default_domain,
             ),
             Expression::Rand(index) => {
-                // The constraint target for random values defaults to the second (auxiliary) trace
-                // segment.
+                // constraint target for random values defaults to the second trace segment.
```
Comment (grjte) [inline ir/src/constraints/graph.rs:111]: nit: I would probably leave this comment as it was, since we still have lots of references to "auxiliary" throughout the code

File: ir/src/constraints/graph.rs [hunk 8]
```diff
@@ -108,13 +108,18 @@ impl AlgebraicGraph {
...
-                let node_index = self.insert_op(Operation::RandomValue(*index));
-                Ok(ExprDetails::new(node_index, AUX_SEGMENT, default_domain))
+                if *index >= symbol_table.num_random_values() as usize {
+                    return Err(SemanticError::IndexOutOfRange(format!(
+                        "Random value index {} is greater than or equal to the total number of random values ({}).", 
+                        index,
+                        symbol_table.num_random_values()
+                    )));
+                }
+                self.insert_random_value(*index, default_domain, AUX_SEGMENT)
```
Comment (grjte) [inline ir/src/constraints/graph.rs:122]: nit: I'd prefer to pass the symbol table to the `insert_random_values` method as well so that the validity check and error can be moved inside that method  (`if *index >= symbol_table.num_random_values()`). That will keep this `insert_expr` method cleaner & match the pattern of the other branches

File: ir/src/constraints/graph.rs [hunk 9]
```diff
@@ -472,6 +483,16 @@ impl AlgebraicGraph {
...
+
+    fn insert_random_value(
+        &mut self,
+        index: usize,
+        domain: ConstraintDomain,
+        trace_segment: u8,
+    ) -> Result<ExprDetails, SemanticError> {
+        let node_index = self.insert_op(Operation::RandomValue(index));
+        Ok(ExprDetails::new(node_index, trace_segment, domain))
+    }
```
Comment (grjte) [inline ir/src/constraints/graph.rs:495]: nit: I'd move this method to just below `insert_constant`. The ordering of the methods has "access" methods after the others.

**Review by grjte — CHANGES_REQUESTED** [review]
Looks great, thank you @Overcastan! I left several comments inline, but it's all small stuff

File: air-script/tests/random_values/random_values.air (hunk 2 above)
Comment (grjte) [reply air-script/tests/random_values/random_values.air:18]: I noticed in the inline `TODO` below that `alphas` won't be supported yet in the IR, so feel free to ignore "nit 2" for now. Maybe we can update this when the other change is made

File: ir/src/constraints/graph.rs [hunk 10]
```diff
@@ -348,6 +353,9 @@ impl AlgebraicGraph {
             IdentifierType::Constant(ConstantType::Scalar(_)) => {
```
Comment (tohrnii) [inline ir/src/constraints/graph.rs:353]: nit: The doc comment of this method should be updated to include random values.

File: ir/src/symbol_table.rs [hunk 11]
```diff
@@ -249,6 +280,16 @@ impl SymbolTable {
...
+                if vector_access.idx() >= *size {
+                    Err(SemanticError::vector_access_out_of_bounds(
+                        vector_access,
+                        *size,
+                    ))
+                } else {
+                    Ok(symbol_type)
+                }
```
Comment (tohrnii) [inline ir/src/symbol_table.rs:291]: nit: We should probably keep the order of comparison similar to other match arms just to be consistent, ie,
```
if vector_access.idx() < *size {
    Ok(symbol_type)
} else {
    Err(SemanticError::vector_access_out_of_bounds(
        vector_access,
        *size,
    ))
}
```

or change the order for the others if you have preference.

File: ir/src/symbol_table.rs [hunk 12]
```diff
@@ -20,6 +20,9 @@ pub(super) enum IdentifierType {
     PeriodicColumn(usize, usize),
     /// an identifier for an integrity variable, containing its name and value
     IntegrityVariable(Variable),
+    /// an identifier for random value, containing its index in the random values array and its
+    /// length if this value is an array. For non-array random values second parameter is always 1.
+    RandomValue(usize, usize),
```
Comment (tohrnii) [inline ir/src/symbol_table.rs:25]: nit: If this can represent an array, should this be called RandomValues instead?

File: ir/src/constraints/graph.rs [hunk 13]
```diff
@@ -348,6 +353,9 @@ impl AlgebraicGraph {
...
                 self.insert_constant(ConstantValue::Scalar(ident.to_string()), domain)
             }
+            IdentifierType::RandomValue(offset, _size) => {
```
Comment (tohrnii) [inline ir/src/constraints/graph.rs:356]: minor nit: @grjte pointed this out elsewhere that it would be better to do RandomValue(offset, _)

File: ir/src/constraints/graph.rs [hunk 14]
```diff
@@ -426,6 +434,9 @@ impl AlgebraicGraph {
             IdentifierType::PublicInput(_) => {
                 unimplemented!("TODO: add support for public inputs.")
             }
+            IdentifierType::RandomValue(offset, _size) => {
```
Comment (tohrnii) [inline ir/src/constraints/graph.rs:437]: minor nit: @grjte pointed this out elsewhere that it would be better to do RandomValue(offset, _)

**Review by tohrnii — CHANGES_REQUESTED** [review]
Thank you @Overcastan, looks good to me. I just left a few small nits inline.

File: ir/src/symbol_table.rs (hunk 12 above)
Comment (Overcastan) [reply ir/src/symbol_table.rs:25]: The problem is that we have a main array of random values (`rand: [a, b, c, ... ]`), so calling it `RandomValues` can be confusing, it will be unclear whether we mean the main array or its element, which is represented by an array. I agree with you that it would be better to rename it somehow, but I don't know yet which name would be better. Perhaps `RandomValuesElement', but this name also does not reflect that this element can be an array.

File: ir/src/symbol_table.rs (hunk 5 above)
Comment (Overcastan) [reply ir/src/symbol_table.rs:168]: Initially, I decided to leave the implementation of this for later, because the PR turned out to be quite large, and I did not want to complicate it with additional optional functionality. But we split it into two, so I don't see any more problems not including it in the current PR. So I'll do it now!

File: ir/src/symbol_table.rs (hunk 12 above)
Comment (grjte) [reply ir/src/symbol_table.rs:25]: It could be `RandomValuesBinding` again, to be somewhat consistent with the AST?

Comment (tohrnii) [reply ir/src/symbol_table.rs]: What about `RandomValuesGroup`?

File: ir/src/tests/mod.rs [hunk 15]
```diff
@@ -379,6 +379,120 @@ fn transition_constraints_using_parens() {
...
+        enf a' = c + 1
+    boundary_constraints:
+        enf a.first = $rand[10] * 2
+        enf a.last = 1";
+
+    let parsed = parse(source_ident_vector_index).expect("Parsing failed");
+    let result = AirIR::from_source(&parsed);
+    assert!(result.is_err());
+}
+
```
Comment (tohrnii) [inline ir/src/tests/mod.rs:495]: I think we should change these so that random values are only used against auxiliary columns. We don't have a check for it yet but we should still have valid tests here for now. Also, do you think we should address #46 here as well or should we do that in another PR?

File: ir/src/symbol_table.rs [hunk 16]
```diff
@@ -151,6 +159,31 @@ impl SymbolTable {
         Ok(())
     }
 
+    /// Adds all random values by their identifier names and array length.
+    pub(super) fn insert_random_values(
+        &mut self,
+        values: &RandomValues,
+    ) -> Result<(), SemanticError> {
+        // TODO: add name of random values vector to the identifiers. For now it can be called only
+        // by `$rand`
```
Comment (tohrnii) [inline ir/src/symbol_table.rs:168]: nit: I think this comment can be removed now?

**Review by tohrnii — CHANGES_REQUESTED** [review]
Thank you @Overcastan, looks good to me. I just left very minor things inline. One question is if we should handle #46 in this PR and the check that you suggested regarding returning an error if there are no aux trace columns but random values section is declared or if we should leave it for the next PR.

File: ir/src/tests/mod.rs (hunk 15 above)
Comment (Overcastan) [reply ir/src/tests/mod.rs:495]: I think I can do it in this PR, looks like it wouldn't involve a lot of changes. Just one question: should we also throw an error if random values are used against the main trace in integrity constraints?

**Comment by Overcastan** [comment]
I really don't like how I implemented additional checks, but my other ideas didn't work (it's too hard to implement them). Does anyone have suggestions how it can be done better?

File: ir/src/tests/mod.rs (hunk 15 above)
Comment (tohrnii) [reply ir/src/tests/mod.rs]: I don't think so. I think we can have integrity constraints with random values and main trace column access but maybe it should be an error if there are random values but no auxiliary column access in a constraint? I'm not sure, maybe @grjte or @bobbinth could confirm?

Comment (bobbinth) [reply ir/src/tests/mod.rs]: Ideally, something like this should be a warning - but since we don't have warnings, I'm not quite sure what's the best approach. Probably having it be an error is better - but if it creates a lot of extra work now, we can do it later too.

File: ir/src/constraints/graph.rs [hunk 17]
```diff
@@ -107,14 +107,19 @@ impl AlgebraicGraph {
...
                 // used. This requires having a way to describe different sets of randomness in
                 // the AirScript syntax.
-                let node_index = self.insert_op(Operation::RandomValue(*index));
-                Ok(ExprDetails::new(node_index, AUX_SEGMENT, default_domain))
+                self.insert_random_access(
+                    symbol_table,
+                    name.name(),
+                    *index,
+                    default_domain,
+                    AUX_SEGMENT,
```
Comment (grjte) [inline ir/src/constraints/graph.rs:121]: nit: Let's keep the order of the domain and segment consistent everywhere. Since we do the segment first when initializing `ExprDetails`, we should do it first here as well

File: ir/src/constraints/graph.rs [hunk 18]
```diff
@@ -273,6 +278,27 @@ impl AlgebraicGraph {
...
     }
 
+    /// Inserts random value with specified index into the graph and returns the resulting
+    /// expression details.
+    fn insert_random_value(
+        &mut self,
+        symbol_table: &SymbolTable,
+        index: usize,
+        domain: ConstraintDomain,
+        trace_segment: u8,
```
Comment (grjte) [inline ir/src/constraints/graph.rs:288]: nit: as noted, let's reverse the order of these

File: ir/src/constraints/graph.rs [hunk 19]
```diff
@@ -472,6 +507,35 @@ impl AlgebraicGraph {
...
+    /// # Example
+    /// This function inserts values like `$alphas[3]`, having `name` as `alphas` and `index` as
+    /// `3`
+    fn insert_random_access(
+        &mut self,
+        symbol_table: &SymbolTable,
+        name: &str,
+        index: usize,
+        domain: ConstraintDomain,
+        trace_segment: u8,
```
Comment (grjte) [inline ir/src/constraints/graph.rs:527]: nit: let's reverse the order of these

File: ir/src/lib.rs [hunk 20]
```diff
@@ -191,4 +196,21 @@ impl AirIR {
...
+                if trace.aux_cols.is_empty() {
+                    return Err(SemanticError::InvalidUsage(
+                        "Random Values section cannot be declared without aux trace columns"
+                            .to_string(),
+                    ));
+                }
+            }
+        }
+        Ok(())
+    }
```
Comment (grjte) [inline ir/src/lib.rs:215]: We probably want to use the source `validator` for this, so lets leave it for a future PR.

File: ir/src/lib.rs [hunk 21]
```diff
@@ -2,6 +2,7 @@ pub use air_script_core::{
     Constant, ConstantType, Expression, Identifier, IndexedTraceAccess, MatrixAccess,
     NamedTraceAccess, TraceSegment, Variable, VariableType, VectorAccess,
 };
+use parser::ast::SourceSection;
```
Comment (grjte) [inline ir/src/lib.rs:5]: It would be better to access `SourceSection` consistently everywhere, so it should either always be `ast::SourceSection` or `SourceSection`. Since we've been using the former one so far, I would stick with that, and then we don't need this import.

File: ir/src/constraints/mod.rs [hunk 22]
```diff
@@ -264,6 +266,11 @@ impl Constraints {
                     domain,
                 )?;
 
+                let contains_ranom_value = self.contains_rand_value(&rhs.root_idx());
+                if contains_ranom_value && lhs.trace_segment() == DEFAULT_SEGMENT {
+                    return Err(SemanticError::InvalidUsage("Random Values cannot be used in boundary constraints defined against main trace columns".to_string()));
+                }
+
```
Comment (grjte) [inline ir/src/constraints/mod.rs:273]: Let's remove this for now and deal with it in a future PR. Then the `DEFAULT_SEGMENT` doesn't need to be `pub` anymore

File: ir/src/symbol_table.rs [hunk 23]
```diff
@@ -151,6 +159,33 @@ impl SymbolTable {
...
+    /// Adds all random values by their identifier names and array length.
+    pub(super) fn insert_random_values(
+        &mut self,
+        values: &RandomValues,
+    ) -> Result<(), SemanticError> {
+        self.num_random_values = values.size() as u16;
+        let mut offset = 0;
+
+        let a = Some(10);
+        if a.is_some() {}
```
Comment (grjte) [inline ir/src/symbol_table.rs:171]: It looks like we can get rid of this

File: ir/src/symbol_table.rs [hunk 24]
```diff
@@ -151,6 +159,33 @@ impl SymbolTable {
...
+
+        // add array name to the symbol table
+        self.insert_symbol(
+            values.name(),
+            IdentifierType::RandomValuesBinding(offset, values.size() as usize),
+        )?;
+        // add array values to the symbol table
```
Comment (grjte) [inline ir/src/symbol_table.rs:178]: nit: I think these comments could be slightly clearer. I'd probably do the following:

```
...
        // add the name of the random values array to the symbol table
        self.insert_symbol(
            values.name(),
            IdentifierType::RandomValuesBinding(offset, values.size() as usize),
        )?;
        // add the named random value bindings to the symbol table
...
```

File: ir/src/tests/mod.rs [hunk 25]
```diff
@@ -379,6 +379,167 @@ fn transition_constraints_using_parens() {
...
+    integrity_constraints:
+        enf c' = $rand[3] + 1
+    boundary_constraints:
+        enf c.first = $rand[10] * 2
+        enf c.last = 1";
+
+    let parsed = parse(source).expect("Parsing failed");
+    let result = AirIR::from_source(&parsed);
+    assert!(result.is_ok());
+}
```
Comment (grjte) [inline ir/src/tests/mod.rs:401]: nit: here and on all tests, I'd prefer to put the `boundary_constraints` before the `integrity_constraints`, since that's generally what we do, and consistency tends to be helpful

File: ir/src/tests/mod.rs [hunk 26]
```diff
@@ -379,6 +379,167 @@ fn transition_constraints_using_parens() {
...
+    integrity_constraints:
+        enf c' = $rand[3] + 1
+    boundary_constraints:
+        enf a.first = $rand[10] * 2
+        enf b[2].last = 1";
+
+    let parsed = parse(source).expect("Parsing failed");
+    let result = AirIR::from_source(&parsed);
+    assert!(result.is_err());
+}
```
Comment (grjte) [inline ir/src/tests/mod.rs:541]: Since we're not going to add this check until later, you could add the `ignore` attribute to this test for now and a `TODO` that we'll turn it on later (or it's fine to just remove the test and we can write it again later)

**Review by grjte — CHANGES_REQUESTED** [review]
Thanks @Overcastan! I left a few more comments inline. I think we can simplify and leave the validity checks for a future PR, since this one is good overall and I'd rather merge it sooner

File: ir/src/symbol_table.rs [hunk 27]
```diff
@@ -32,6 +35,7 @@ impl Display for IdentifierType {
                 write!(f, "TraceColumns in segment {}", columns.trace_segment())
             }
             Self::IntegrityVariable(_) => write!(f, "IntegrityVariable"),
+            Self::RandomValuesBinding(_, _) => write!(f, "RandomValue"),
```
Comment (tohrnii) [inline ir/src/symbol_table.rs:38]: minor nit: Should this be RandomValuesBinding?

File: ir/src/tests/mod.rs [hunk 28]
```diff
@@ -379,6 +379,126 @@ fn transition_constraints_using_parens() {
     assert!(result.is_ok());
 }
 
+#[test]
+fn random_values_fixed_list() {
```
Comment (tohrnii) [inline ir/src/tests/mod.rs:383]: nit: Should this be renamed to `random_values_indexed_access` or something like that?

File: ir/src/tests/mod.rs [hunk 29]
```diff
@@ -379,6 +379,126 @@ fn transition_constraints_using_parens() {
...
+    integrity_constraints:
+        enf c' = $alphas[3] + 1";
+
+    let parsed = parse(source).expect("Parsing failed");
+    let result = AirIR::from_source(&parsed);
+    assert!(result.is_ok());
+}
+
+#[test]
+fn random_values_ident_vector() {
```
Comment (tohrnii) [inline ir/src/tests/mod.rs:425]: nit: Should this be renamed to `random_values_named_access` or something similar?

**Review by tohrnii — APPROVED** [review]
Thank you @Overcastan, looks good to me. I just left a few very minor naming related nits but don't have strong opinion on those.

File: ir/src/symbol_table.rs (hunk 27 above)
Comment (Overcastan) [reply ir/src/symbol_table.rs:38]: My bad, I overlooked it. Thank you!

**Review by grjte — APPROVED** [review]
LGTM! Thanks @Overcastan!
